File: easy/24. Generate Document/generate_document.cpp
#include <string>
#include <unordered_map>

bool generateDocument(std::string characters, std::string document) {
  if (document.length() == 0) {
    return true;
  }
  std::unordered_map<char, int> charCount;
  for (char c : characters) {
    if (charCount.find(c) == charCount.end()) {
      charCount[c] = 0;
    }
    charCount[c]++;
  }
  for (int i = 0; i < document.size(); i++) {
    if (charCount.find(document[i]) == charCount.end() ||
        charCount[document[i]] == 0) {
      return false;
    }
    charCount[document[i]]--;
  }
  return true;
}

File: dsa/easy/Generate Document/generate_document.go
package main

func GenerateDocument(characters string, document string) bool {
	if len(document) == 0 {
		return true
	}
	var charMap = make(map[rune]int)
	for _, char := range characters {
		if _, ok := charMap[char]; ok {
			charMap[char]++
		} else {
			charMap[char] = 1
		}
	}
	for _, char := range document {
		if count, ok := charMap[char]; !ok || count == 0 {
			return false
		}
		charMap[char]--
	}

	return true
}
